pkg/controller/synchronize.go

```diff
@@ -0,0 +1,93 @@
+package controller
+
+import (
+	"fmt"
+	"github.com/oslokommune/okctl/pkg/controller/reconsiler"
+	"github.com/oslokommune/okctl/pkg/controller/resourcetree"
+	"github.com/spf13/afero"
+)
+
+type SynchronizeOpts struct {
+	DesiredTree *resourcetree.ResourceNode
+	
+	ReconsiliationManager *reconsiler.ReconsilerManager
+	
+	Fs *afero.Afero
+	OutputDir string
+
+	GithubGetter reconsiler.GithubGetter
+	GithubSetter reconsiler.GithubSetter
+
+	CIDRGetter StringFetcher
+	PrimaryHostedZoneGetter HostedZoneFetcher
+}
+
+// Synchronize knows how to discover differences between desired and actual state and rectify them
+func Synchronize(opts *SynchronizeOpts) error {
+	opts.DesiredTree.SetStateRefresher(resourcetree.ResourceNodeTypeCluster, CreateClusterStateRefresher(
+		opts.Fs,
+		opts.OutputDir,
+		opts.CIDRGetter,
+	))
+	
+	opts.DesiredTree.SetStateRefresher(resourcetree.ResourceNodeTypeALBIngress, CreateALBIngressControllerRefresher(
+		opts.Fs,
+		opts.OutputDir,
+	))
+
+	opts.DesiredTree.SetStateRefresher(resourcetree.ResourceNodeTypeExternalDNS, CreateExternalDNSStateRefresher(
+		opts.PrimaryHostedZoneGetter,
+	))
+	
+	opts.DesiredTree.SetStateRefresher(resourcetree.ResourceNodeTypeIdentityManager, CreateIdentityManagerRefresher(
+		opts.PrimaryHostedZoneGetter,
+	))
+
+	opts.DesiredTree.SetStateRefresher(resourcetree.ResourceNodeTypeGithub, CreateGithubStateRefresher(
+		opts.GithubGetter,
+		opts.GithubSetter,
+	))
+	
+	opts.DesiredTree.SetStateRefresher(resourcetree.ResourceNodeTypeArgoCD, CreateArgocdStateRefresher(
+		opts.PrimaryHostedZoneGetter,
+	))
+
+	currentStateGraphOpts, err := NewCreateCurrentStateGraphOpts(opts.Fs, opts.OutputDir, opts.GithubGetter)
+	if err != nil {
+	    return fmt.Errorf("unable to get existing services: %w", err)
+	}
+
+	currentStateTree := CreateCurrentStateGraph(currentStateGraphOpts)
+
+	diffGraph := *opts.DesiredTree
+
+	diffGraph.ApplyFunction(applyCurrentState, currentStateTree)
+
+	return handleNode(opts.ReconsiliationManager, &diffGraph)
+}
+
+// handleNode knows how to run Reconsile() on every node of a ResourceNode tree
+func handleNode(reconsilerManager *reconsiler.ReconsilerManager, currentNode *resourcetree.ResourceNode) error {
+	_, err := reconsilerManager.Reconsile(currentNode)
```

Comment:
Are we ignoring the ReconsilationResult because this being a MVP? What are possible uses of it in the future?

Reply:
Ugh. This is me future proofing. 

However, the reconciliation result will be necessary when implementing the nameserver dance. I plan on creating a node which ensures the nameservers is in place - and this node will have to have ReconciliationResult{requeue: true} until the node sees that the nameservers has been added to the oslo.systems domain